### ISSUE TITLE
Incorrect version in some distribution packages

### ISSUE THREAD

**Issue by wader** [wader/fq#68]
```sh
$ docker run --rm alpine:3.15 sh -c 'apk add -X http://dl-cdn.alpinelinux.org/alpine/edge/testing fq && fq --version'
fetch http://dl-cdn.alpinelinux.org/alpine/edge/testing/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/community/x86_64/APKINDEX.tar.gz
(1/1) Installing fq (0.0.3-r0)
Executing busybox-1.34.1-r3.trigger
OK: 15 MiB in 15 packages
dev
```

Currently version is set by goreleaser/make when building using go build ldflag `-X main.version=...`. To make it work with distributions that builds manually using `go build ` etc we probably have to include the version in the source somehow.

- Create some version file using goreleaser when packaging source archives. Wont fix issue if distribution clones the repo.
- Convince package maintainers to build using make and pass version as some argument? too much hassel i think
- Manually change/commit some version file before release
- Some other way?

**Comment by Doctor-love** [comment]
Hmm, there is no way to include this information using standard "go build"?

**Comment by wader** [comment]
There is but not automatically, in go 1.18 go build will start to include some VCS information automatically i think. But that wont help for source archive builds. For binaries I currently i use goreleaser to pass go build flags to set the variable `main.version`.

I see that the nix builds sets verison https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/fq/default.nix#L24
but alpine does not https://git.alpinelinux.org/aports/tree/testing/fq/APKBUILD

I don't really follow how nix, alpine etc downloads the source, clones a git tag or downloads release archive?

I'll think and investigate a bit more, maybe can make goreleaser generate a version file while arching? but feels a bit ugly

**Comment by wader** [comment]
@mpanczyk hey, what do you think about this?

**Comment by mpanczyk** [comment]
@wader Sorry, but I think you have me confused with someone else.

**Comment by wader** [comment]
@mpanczyk aha sorry! thought you might be the maintainer of this package https://pkgs.alpinelinux.org/package/edge/testing/x86_64/fq

**Comment by wader** [comment]
gojq does this by keeping the latest released in a go file https://github.com/itchyny/gojq/blob/main/cli/cli.go#L19 and also has a "revision" variable set to "HEAD" by default and is changed it building via make. Feel like a good solution i think.

**Comment by mpolanski** [comment]
This is now fixed on Alpine via https://github.com/alpinelinux/aports/commit/1974aa4cc61f645636e3c3347ef0294334d9d732. I don't think that you need to change how the version string is set. A lot of software use goreleaser and pass some flags using ldflags. It's nothing wrong with that.

**Comment by wader** [comment]
@mpolanski  👍  ill let you know if i switch to baking the version number into the source somehow, might be nice for ppl who build from source archives

**Comment by wader** [comment]
@danfe Looking at the freebsd ports Makefile i guess it also shows "dev" as version? if so, any preference to how to fix?

**Comment by danfe** [comment]
>  Looking at the FreeBSD ports `Makefile` I guess it also shows `dev` as version?

That's right, it does show `dev` when running `fq --version`.

> If so, any preference to how to fix?

This is one of those things we lost upon entering modern era which replaced traditional, properly bootstrapped and `$version` expanded release tarballs with git tags (git does not even support keyword expansion which makes it even harder).  Archived sources from GitHub (which packagers download, it's not the same as compressed `git clone`) do not contain any git metadata, so there's nowhere to derive version from (ditto for the commit hash).  I personally simply edit files with `sed(1)` to fill in correct values prior to building.  I can't think of an elegant solution (with git) other than generating proper release tarball like in the old times.

**Comment by wader** [comment]
> This is one of those things we lost upon entering modern era which replaced traditional, properly bootstrapped and `$version` expanded release tarballs with git tags (git does not even support keyword expansion which makes it even harder). Archived sources from GitHub (which packagers download, it's not the same as compressed `git clone`) do not contain any git metadata, so there's nowhere to derive version from (ditto for the commit hash). I personally simply edit files with `sed(1)` to fill in correct values prior to building. I can't think of an elegant solution (with git) other than generating proper release tarball like in the old times.

Yeah i think your right. I will probably keep the latest release version in a file somehow. Also noticed now that goreleaser (a go release tool) seems to not produce source archives on it own, instead links to archive tags somehow, should look into that.

Thanks for your thoughts

**Comment by wader** [comment]
@danfe @mpanczyk Hey, just published version 0.0.4 and embeds the version so build flags is no needed anymore.
